tests/linearizability/model.go

```diff
@@ -93,121 +91,83 @@ var etcdModel = porcupine.Model{
+		state.Key = request.Key
+		ok, v := initValueRevision(request, response)
+		if ok {
+			state.PossibleValues = append(state.PossibleValues, v)
+		}
+		return true, state
 	}
 	if state.Key != request.Key {
-		panic("Multiple keys not supported")
+		panic("multiple keys not supported")
 	}
-	switch request.Op {
-	case Get:
-		return stepGet(state, request, response)
-	case Put:
-		return stepPut(state, request, response)
-	case Delete:
-		return stepDelete(state, request, response)
-	default:
-		panic("Unknown operation")
+	if response.Err != nil {
+		for _, kv := range state.PossibleValues {
+			kv.handle(request)
+			state.PossibleValues = append(state.PossibleValues, kv)
+		}
+	} else {
+		var i = 0
+		for _, kv := range state.PossibleValues {
+			expectedResponse := kv.handle(request)
+			if expectedResponse == response {
+				state.PossibleValues[i] = kv
+				i++
+			}
+		}
+		state.PossibleValues = state.PossibleValues[:i]
 	}
+	return len(state.PossibleValues) > 0, state
+}
+
+type valueRevision struct {
+	Value    string
+	Revision int64
 }
 
-func initState(request EtcdRequest, response EtcdResponse) EtcdState {
-	state := EtcdState{
-		Key:          request.Key,
-		LastRevision: response.Revision,
+func initValueRevision(request EtcdRequest, response EtcdResponse) (ok bool, v valueRevision) {
+	if response.Err != nil {
+		return false, valueRevision{}
 	}
 	switch request.Op {
 	case Get:
-		state.Value = response.GetData
+		return true, valueRevision{
+			Value:    response.GetData,
+			Revision: response.Revision,
+		}
 	case Put:
-		if response.Err == nil {
-			state.Value = request.PutData
-		} else {
-			state.FailedWrite = &request
+		return true, valueRevision{
+			Value:    request.PutData,
+			Revision: response.Revision,
 		}
 	case Delete:
-		if response.Err != nil {
-			state.FailedWrite = &request
+		return true, valueRevision{
+			Value:    "",
+			Revision: response.Revision,
 		}
 	default:
 		panic("Unknown operation")
 	}
-	return state
 }
 
-func stepGet(state EtcdState, request EtcdRequest, response EtcdResponse) (bool, EtcdState) {
-	if state.Value == response.GetData && state.LastRevision == response.Revision {
-		state.FailedWrite = nil
-		return true, state
-	}
-	if state.FailedWrite != nil && state.LastRevision < response.Revision {
-		var ok bool
-		switch state.FailedWrite.Op {
-		case Get:
-			panic("Expected write")
-		case Put:
-			ok = response.GetData == state.FailedWrite.PutData
-		case Delete:
-			ok = response.GetData == ""
-		default:
-			panic("Unknown operation")
-		}
-		if ok {
-			state.Value = response.GetData
-			state.LastRevision = response.Revision
-			state.FailedWrite = nil
-			return true, state
+func (v *valueRevision) handle(request EtcdRequest) EtcdResponse {
+	response := EtcdResponse{}
+	switch request.Op {
+	case Get:
+		response.GetData = v.Value
+	case Put:
+		v.Value = request.PutData
+		v.Revision += 1
+	case Delete:
+		if v.Value != "" {
+			v.Value = ""
+			v.Revision += 1
```

Comment:
It might not be correct to in-place update the history?  Otherwise, you need to differentiate successful and failed request? If I understand correctly, this method is just to generate an expected response based on current state.

Reply:
I wrote the code in a way that it is. When iterating through possible values I always copy.
This is tested by all pre-existing tests.

Reply:
If you prefer that we don't update `ValueRevision` in place for readability please let me know. I compared both versions and picked this one.

Reply:
This is not just readability, it also is wrong.  The receiver is pointer, any in-place updating will affect the caller. The reason why you do not run into issue is because the golang for--range loop made a copy for you. It's a potential risk that new contributors may misunderstand and generate buggy code. 

Please do not modify the value in-place.

Reply:
> The reason why you do not run into issue is because the golang for--range loop made a copy for you.

Yes, I know, That was intentional so the code was correct. Rewrote to copy for clarity.